components/salsa-macro-rules/src/setup_input_struct.rs

```diff
@@ -121,14 +127,33 @@ macro_rules! setup_input_struct {
             }
 
             impl $Struct {
-                pub fn $new_fn<$Db>(db: &$Db, $($field_id: $field_ty),*) -> Self
+                #[inline]
+                pub fn $new_fn<$Db>(db: &$Db, $($required_field_id: $required_field_ty),*) -> Self
                 where
                     // FIXME(rust-lang/rust#65991): The `db` argument *should* have the type `dyn Database`
                     $Db: ?Sized + salsa::Database,
                 {
-                    let current_revision = $zalsa::current_revision(db);
-                    let stamps = $zalsa::Array::new([$zalsa::stamp(current_revision, Default::default()); $N]);
-                    $Configuration::ingredient(db.as_salsa_database()).new_input(($($field_id,)*), stamps)
+                    Self::builder($($required_field_id,)*).new(db)
+                }
+
+                pub fn builder($($required_field_id: $required_field_ty),*) -> <Self as $zalsa_struct::HasBuilder>::Builder
```

Comment:
It would be nice if we could avoid the extra `required_fields` macro argument but I couldn't figure out how to use `macro_if` in a function declaration.

Reply:
yeah, you'd have to wrap the whole function. seems fine this way.